grc/core/params/param.py

```diff
@@ -154,9 +154,24 @@ def validate(self):
             self.add_error_message('Type "{}" is not a possible type.'.format(self.dtype))
 
         validator = dtypes.validators.get(self.dtype, None)
+        #build  dynyamic blacklist from imports if needed
+        bl = {}
+        if self.dtype == 'id':
+            importlist=self.parent_flowgraph.imports()
+            for expr in importlist:
+                try:
+                    exec(expr, bl)
+                except ImportError:
+                    # We do not have a good way right now to determine if an import is for a
+                    # hier block, these imports will fail as they are not in the search path
+                    # this is ok behavior, unfortunately we could be hiding other import bugs
+                    pass
+                except Exception:
+                    log.exception('Failed to evaluate import expression "{0}"'.format(expr), exc_info=True)
+                    pass
```

Comment:
This re-creates the black list for every block - even though it does not change. My suggestion here is to have the FlowGraph class build that list as it creates the namespace (and runs the imports anyway). Then have getter, `get_imported_names()` which you can use below

Reply:
Many thanks for your hints. This makes the code smoother